flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/deserializer/KafkaRecordDeserializationSchema.java

```diff
@@ -123,9 +121,8 @@ default void open(DeserializationSchema.InitializationContext context) throws Ex
      * @return A {@link KafkaRecordDeserializationSchema} that deserialize the value with the given
      *     deserializer.
      */
-    static <V, D extends Configurable & Deserializer<V>>
-            KafkaRecordDeserializationSchema<V> valueOnly(
-                    Class<D> valueDeserializerClass, Map<String, String> config) {
+    static <V, D extends Deserializer<V>> KafkaRecordDeserializationSchema<V> valueOnly(
```

Comment:
Prior to https://github.com/apache/flink/pull/15051, we have the following method in the class `KafkaRecordDeserializer`:

```
    /**
     * Wraps a Kafka {@link Deserializer} to a {@link KafkaRecordDeserializer}.
     *
     * @param valueDeserializerClass the deserializer class used to deserialize the value.
     * @param config the configuration of the value deserializer, only valid when the deserializer
     *     is an implementation of {@code Configurable}.
     * @param <V> the value type.
     * @param <D> the type of the deserializer.
     * @return A {@link KafkaRecordDeserializer} that deserialize the value with the given
     *     deserializer.
     */
    static <V, D extends Configurable & Deserializer<V>> KafkaRecordDeserializer<V> valueOnly(
            Class<D> valueDeserializerClass, Map<String, String> config) {
    ...
    }
```

This method basically assumes that users might have defined a deserializerA as a subclass of both Configurable and Deserializer, where `deserializerA::configure(Map<String, ?>)` could have user-defined configuration logic. Since this subclass could be defined by user outside the Flink project, we want to continue to support this this user-defined deserializer for backward compatibility.

Changing the method signature in this way could cause backward incompatibility. Though this method is not marked `@public`, I think this is because we forgot to add this tag. Given that this method is supposed to be used by public users, I feel it is preferred not to break it.

What do you think?

Reply:
BTW, this method (prior to https://github.com/apache/flink/pull/15051) was added in FLINK-18323. Jiangjie Qin, author of that PR, was pretty familiar with Kafka. I believe he has good reason to add that method, which uses `org.apache.kafka.common.Configurable.configure` instead of `org.apache.kafka.common.serialization.Deserializer.configure`. And I guess the reason is likely due to backward compatibility.

Reply:
BTW, this method (prior to https://github.com/apache/flink/pull/15051) was added in FLINK-18323. Jiangjie Qin, author of that PR, was pretty familiar with Kafka. I believe he has good reason to add that method, which uses `org.apache.kafka.common.Configurable.configure` instead of `org.apache.kafka.common.serialization.Deserializer.configure`. And I guess the reason is likely due to backward compatibility.

Reply:
Thanks for the review. 

Backward compatibility has been considered. This PR just reduced the type control of the input parameter DeserializationSchema a little bit from  `Configurable & Deserializer<V>` to `Deserializer<V>`. The `[Wrapper](https://github.com/apache/flink/pull/18397/files#diff-ed43df5179d747c55bc2b9478ead0b15d49a5275e76827121d31904db779449fL68)` will check internally and call `org.apache.kafka.common.Configurable.configure`  if the schema implements `Configurable`.

Reply:
@lindong28 thanks for your feedback. I am wondering a bit what the best way here is. Removing the method seems not ideal because as you said users might have used it with Serializers/Deserializers implementing Configurable.

In the Kafka ecosystem there seem to be different ways of configuring a Serializer/Deserializer one is the current one with implementing Configurable and the other is through the Serializer/Deserializer itself [1]. I am no Kafka expert so I don't know the right way to configure them. I guess we can also support both methods.

[1] https://github.com/apache/kafka/blob/3dd3f3ef035501816efac23c32d517bf62fdaa56/clients/src/main/java/org/apache/kafka/common/serialization/Serializer.java#L40

Reply:
@JingGe I see. I agree it is still backward compatible (suppose we remove the `config.isEmpty()` check) if user invokes this method with a `Configurable` instance.

It seems that the Java doc of this method still says `only valid when the deserializer is an implementation of {@code Configurable}`. Do we need to update this?

Reply:
@fapaul I think the best way is to use `Deserializer::configure(...)` to configure deserializer. And I believe all the deserializers provided off-the-shelf by Apache Kafka, and most user-implemented deserializer, use `Deserializer::configure(...)` instead of `Configurable`.

`Configurable::configure(...)` is not deprecated in Kafka and I believe Flink still supports this for backward compatibility. And I agree it is useful to extend the functionality of this method to support `Deserializer::configure(...)` as long as we could make sure we don't break users existing application.